cvmfs/cache_ram.h

```diff
@@ -13,75 +13,275 @@
+   * @param size The total size of the new cache entry
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
   virtual int StartTxn(const shash::Any &id, uint64_t size, void *txn);
+
+  /**
+   * Set the transaction parameters. At present, only @ref type is used for pinned, volatile, etc.
+   * @param description Unused
+   * @param type The type of the entry, e.g. pinned
+   * @param flags Unused
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
   virtual void CtrlTxn(const std::string &description,
                        const ObjectType type,
                        const int flags,
                        void *txn);
+
+  /**
+   * Copy the given memory region into the transaction buffer. Copying starts at the transaction's current offset
+   * @param buf The source address
+   * @param size The number of bytes to copy
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
   virtual int64_t Write(const void *buf, uint64_t size, void *txn);
+
+  /**
+   * Seek to the beginning of the transaction buffer
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
   virtual int Reset(void *txn);
+
+  /**
+   * Commit a transaction and open the resulting cache entry. This is necessary to avoid a race condition in which the
+   * cache entry is evicted between the calls to CommitTxn and Open.
+   * @param txn A pointer to space allocated for storing the transaction details
+   * @returns A file descriptor to the new cache entry, or a negative error from @ref CommitTxn
+   */
   virtual int OpenFromTxn(void *txn);
+
+  /**
+   * Free the resources allocated for the pending transaction
+   * @param txn A pointer to space allocated for storing the transaction details
+   */
   virtual int AbortTxn(void *txn);
+
+   /**
+   * Commit a transaction to the cache. If there is not enough free space in the cache, first try to make room by evicting
+   * volatile entries. If there is still not enough room, try evicting regular entries. If there is *still* not enough
+   * space, give up an return failure.
+   * @param txn A pointer to space allocated for storing the transaction details
+   * @returns -ENOSPC if the transaction would exceed the size of the cache, -ENFILE if no handles are available,
+   * or -EEXIST if an entry with nonzero reference count already exists
+   */
   virtual int CommitTxn(void *txn);
 
  private:
   struct ReadOnlyFd {
-    ReadOnlyFd() : handle(-1), size(0) { }
-    ReadOnlyFd(int64_t h, uint64_t s) : handle(h), size(s) { }
-    /**
-     * Negative handle: invalid FD
-     */
-    int64_t handle;
-    uint64_t size;
+    ReadOnlyFd() : handle(shash::Any()), pos(0), store(NULL) { }
+    ReadOnlyFd(const shash::Any &h, uint64_t pos)
+      : handle(h)
+      , pos(pos)
+      , store(NULL) { }
+    shash::Any handle;
+    uint64_t pos;
+    kvstore::MemoryKvStore *store;
```

Comment:
Do we need to store the `kvstore` in the `ReadOnlyFd`?  Could we just store it once in the `RamCacheManager` object?